api.go

```diff
@@ -146,73 +154,152 @@ type Raft struct {
+	// Validate the Raft server config.
+	if err := ValidateConfig(conf); err != nil {
+		return err
+	}
+
+	// Sanity check the Raft peer configuration.
 	if err := checkConfiguration(configuration); err != nil {
 		return err
 	}
 
-	// Make sure we don't have a current term
+	// Make sure the cluster is in a clean state.
+	bs, err := IsClusterBootstrapped(logs, stable, snaps)
+	if err != nil {
+		return err
+	}
+	if bs {
+		return fmt.Errorf("bootstrap only works on new clusters, cluster has been bootstrapped previously")
+	}
+
+	// Set current term to 1.
+	if err := stable.SetUint64(keyCurrentTerm, 1); err != nil {
+		return fmt.Errorf("failed to save current term: %v", err)
+	}
+
+	// Append configuration entry to log.
+	entry := &Log{
+		Index: 1,
+		Term:  1,
+	}
+	if conf.ProtocolVersion < 2 {
+		entry.Type = LogRemovePeerDeprecated
+		entry.Data = encodePeers(configuration, trans)
+	} else {
+		entry.Type = LogConfiguration
+		entry.Data = encodeConfiguration(configuration)
+	}
+	if err := logs.StoreLog(entry); err != nil {
+		return fmt.Errorf("failed to append configuration entry to log: %v", err)
+	}
+
+	return nil
+}
+
+// RecoverCluster is used to manually force a new configuration in order to
+// recover from a loss of quorum where the current configuration cannot be
+// restored, such as when several servers die at the same time. If this is used,
+// ALL SERVERS in the cluster must be started with the same recovery settings.
+// Otherwise, the configuration may be incorrect depending on which server gets
+// elected.
+func RecoverCluster(conf *Config, logs LogStore, trans Transport,
+	configuration Configuration) error {
+	// Validate the Raft server config.
+	if err := ValidateConfig(conf); err != nil {
+		return err
+	}
+
+	// Sanity check the Raft peer configuration.
+	if err := checkConfiguration(configuration); err != nil {
+		return err
+	}
+
+	// Read the last log value.
+	lastIdx, err := logs.LastIndex()
+	if err != nil {
+		return fmt.Errorf("failed to find last log: %v", err)
+	}
+
+	// Get the log.
+	var lastLog Log
+	if lastIdx > 0 {
+		if err = logs.GetLog(lastIdx, &lastLog); err != nil {
+			return fmt.Errorf("failed to get last log: %v", err)
+		}
+	} else {
+		return fmt.Errorf("Recover only works on non-empty logs")
+	}
+
+	// Add a new log entry.
+	fakeIndex := lastLog.Index + 1
+	entry := &Log{
+		Index: fakeIndex,
+		Term:  lastLog.Term,
+	}
+	if conf.ProtocolVersion < 2 {
+		entry.Type = LogRemovePeerDeprecated
+		entry.Data = encodePeers(configuration, trans)
+	} else {
+		entry.Type = LogConfiguration
+		entry.Data = encodeConfiguration(configuration)
+	}
+	if err := logs.StoreLog(entry); err != nil {
+		return fmt.Errorf("failed to append configuration entry to log: %v", err)
+	}
+
+	return nil
+}
+
+// IsClusterBootstrapped returns true if the cluster has been bootstrapped
+// before.
+func IsClusterBootstrapped(logs LogStore, stable StableStore, snaps SnapshotStore) (bool, error) {
```

Comment:
This is a misleading name (maybe I started it). You can't tell if the cluster has been bootstrapped. At best you can tell if this server has any state.